index.bs

```diff
@@ -249,9 +259,39 @@ If the algorithm using these shorthands is running <a>in parallel</a>, the short
-
-This phrase is useful when you wish to perform multiple asynchronous operations in parallel that return promises, and then react to them all together. If all of the given promises fulfill, then the resulting promise will be fulfilled with an array corresponding to the fulfillment values. If any of them reject, then the resulting promise will be rejected with the first rejection reason to occur.
+To <dfn id="waiting-for-all" export lt="wait for all|waiting for all">wait for all</dfn> of a <a>list</a> of promises |promises|, with success steps |successSteps| that take a <a>list</a> of JavaScript values and failure steps |failureSteps| that take a rejection reason JavaScript value, perform the following steps:
+
+1. Let |rejectionHandler| be a built-in function that takes an argument |arg| and performs the following steps:
+    1. Perform |failureSteps| given |arg|.
+1. Let |index| be 0.
+1. Let |fullfilledCount| be 0.
+1. Let |total| be |promises|'s [=list/size=].
+1. Let |result| be a <a>list</a> containing |total| null values.
+1. [=list/For each=] |promise| of |promises|:
+    1. Let |promiseIndex| be |index|.
+    1. Let |fulfillmentHandler| be a built-in function that takes an argument |arg| performs the following steps:
+        1. Set |result|[|promiseIndex|] to |arg|.
+        1. Set |fullfilledCount| to |fullfilledCount| + 1.
+        1. If |fullfilledCount| equals |total|, then perform |successSteps| given |result|.
+    1. Perform <a abstract-op>PerformPromiseThen</a>(|promise|, |fulfillmentHandler|, |rejectionHandler|).
+    1. Set |index| to |index| + 1.
+
+This phrase is useful when you wish to aggregate the result of multiple promises, and react to them all together, in the same way that {{Promise/all()|Promise.all()}} functions for JavaScript code.
+
+To <dfn id="waiting-for-all-promise" export lt="get a promise to wait for all|getting a promise to wait for all">get a promise for waiting for all</dfn> of a <a>list</a> of promises |promises|, with success steps |successSteps| that take a <a>list</a> of JavaScript values and optional failure steps |failureSteps| that take a rejection reason JavaScript value, perform the following steps:
+
+1. Let |promise| be <a>a new promise</a>.
+1. If |failureSteps| were not given, let them be steps taking an argument |arg| that throw |arg|.
+1. Let |successStepsWrapper| be the following steps, given |results|:
+    1. Let |stepsResult| be the result of performing |successSteps| given |results|. If these steps threw an exception, <a>reject</a> |promise| with that exception.
+    1. <a>Resolve</a> |promise| with |stepsResult|.
+1. Let |failureStepsWrapper| be the following steps, given |reason|:
+    1. Let |stepsResult| be the result of performing |failureSteps| given |reason|. If these steps threw an exception, <a>reject</a> |promise| with that exception.
+    1. <a>Resolve</a> |promise| with |stepsResult|.
```

Comment:
Reject?

Reply:
Nah. If failureSteps return something, we resolve with that thing. If they throw we reject on the previous line.